### PR TITLE
openai-whisper-cpp: init at 1.0.4

### PR DESCRIPTION
**Pull request by dit7ya**
###### Description of changes

https://github.com/ggerganov/whisper.cpp

This is a C++ port of OpenAI's whisper speech-recognition software.

We already have the original whisper package in nixpkgs.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/audio/openai-whisper-cpp/default.nix [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    repo = pname;
+    rev = "3996ecc156486fb93ff505c01090d13192e72aa2";
+    sha256 = "sha256-hPRwjyARaJIjW6YSC7PdPgyao96lSHd5JGEtBLLM+Rg=";
+  };
+
+  patches = [ ./download-models.patch ];
+
+  nativeBuildInputs = [ makeWrapper ];
+
+  buildInputs = [ SDL2 ];
```
Comment (dbaynard) [inline pkgs/tools/audio/openai-whisper-cpp/default.nix:24]: One of these (probably `buildInputs`) is going to need `darwin.apple_sdk.frameworks.Accelerate`, and possibly other frameworks, too. I can't test until the end of the week, though.

File: pkgs/tools/audio/openai-whisper-cpp/default.nix [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  version = "unstable-2022-12-07";
+
+  src = fetchFromGitHub {
+    owner = "ggerganov";
+    repo = pname;
+    rev = "3996ecc156486fb93ff505c01090d13192e72aa2";
+    sha256 = "sha256-hPRwjyARaJIjW6YSC7PdPgyao96lSHd5JGEtBLLM+Rg=";
+  };
+
+  patches = [ ./download-models.patch ];
```
[the remark this answers is not in the document]
Comment (dit7ya) [reply pkgs/tools/audio/openai-whisper-cpp/default.nix:21]: Thanks. Done.

File: pkgs/tools/audio/openai-whisper-cpp/default.nix (hunk 1 above)
Comment (dbaynard) [reply pkgs/tools/audio/openai-whisper-cpp/default.nix]: FYI you need to add `darwin` as the function argument for the expression in the file, and then add the `Accelerate` value as I described, above.

With your change, I get this message.

```console
> nix build -L 'github:dit7ya/nixpkgs/openai-whisper-cpp#openai-whisper-cpp'
error: Function called without required argument "Accelerate" at /nix/store/946xw29yrsryccz32d1k9m74h4315awr-source/pkgs/tools/audio/openai-whisper-cpp/default.nix:7
```

As I said: if nobody else comes to it, I can take a look, myself, at the weekend, but before then I can't make time to do anything other than a few test runs.

**Comment by dbaynard** [comment]
As of 855d07c this builds on aarch64-darwin

**Comment by RaitoBezarius** [comment]
Result of `nixpkgs-review pr 205008` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>darwin.builder</li>
    <li>openai-whisper-cpp</li>
  </ul>
</details>

**Review by RaitoBezarius — APPROVED** [review]

**Comment by RaitoBezarius** [comment]
Did someone try on aarch64-darwin (the platform which is going to interest a lot of people I guess)?

File: pkgs/tools/audio/openai-whisper-cpp/default.nix [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+  nativeBuildInputs = [ makeWrapper ];
+
+  buildInputs = [ SDL2 ] ++ lib.optionals stdenv.isDarwin [ Accelerate CoreGraphics CoreVideo ];
+
+  buildPhase = ''
+    make main
+    make stream
+  '';
```
Comment (SuperSandro2000) [inline pkgs/tools/audio/openai-whisper-cpp/default.nix:36]: Why not use makeFlags?

Comment (RaitoBezarius) [reply pkgs/tools/audio/openai-whisper-cpp/default.nix:36]: also, `runHook preBuild` + `runHook postBuild` (thanks for mentioning this Sandro!)

File: pkgs/tools/audio/openai-whisper-cpp/default.nix [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+  installPhase = ''
+    mkdir -p $out/bin
```
Comment (RaitoBezarius) [inline pkgs/tools/audio/openai-whisper-cpp/default.nix:39]: runHook preInstall / runHook postInstall

**Comment by Builditluc** [comment]
Result of `nixpkgs-review pr 205008` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)

**Review by RaitoBezarius — APPROVED** [review]
LGTM. Did someone try on aarch64-darwin?

**Comment by dbaynard** [comment]
> LGTM. Did someone try on aarch64-darwin?

I can check the build, again. Is there a suitable test to ensure it works correctly?

**Comment by dbaynard** [comment]
2e078c2 builds on aarch64-darwin. And it runs, too (it is _so_ much better than whisper… yeesh — whisper in nixpkgs uses release pytorch, which doesn't support the M1 GPU).

Great job @di7tya

It would be great to include [whisper.cpp/convert-pt-to-ggml.py at master · ggerganov/whisper.cpp](https://github.com/ggerganov/whisper.cpp/blob/ac521a566ea6a79ba968c30101140db9f65d187b/models/convert-pt-to-ggml.py) in the next version.

File: pkgs/tools/audio/openai-whisper-cpp/default.nix [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+, SDL2
+, makeWrapper
+, wget
+, Accelerate
+, CoreGraphics
+, CoreVideo
+}:
+
+stdenv.mkDerivation rec {
+  pname = "whisper.cpp";
```
Comment (dbaynard) [inline pkgs/tools/audio/openai-whisper-cpp/default.nix:13]: Ah, there's one small thing, here — you name the binary `whisper-cpp`, not `whisper.cpp`, meaning `nix run` doesn't just work. I suspect the pname should be `whisper-cpp` (and then you'll have to change the `repo` to be `whisper.cpp`.

**Pull request merged**